File: 142. Linked List Cycle II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

########################### Method 1 ###################################

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        node_set = set()
        while head:
            if head in node_set:
                return head
            node_set.add(head)
            head = head.next
        return None
        

########################### Method 2 ###################################
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        fast = head
        slow = head
        meet = None
        
        while fast:
            slow = slow.next
            fast = fast.next
            if fast == None:
                return None
            fast = fast.next
            if fast == slow:
                meet = fast
                break
                
        while meet and head:
            if meet == head:
                return head
            meet = meet.next
            head = head.next
        return None

File: 142. Linked List Cycle II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 
 **************************** Method 1 ********************************
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
    std::set<ListNode*> node_set;
        while(head){
        if(node_set.find(head)!=node_set.end()){
            return head;
        }
        node_set.insert(head);
        head = head->next;   
        }
        return NULL;
    }
};

**************************** Method 2 ********************************
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *fast = head;
        ListNode *slow = head;
        ListNode *meet = NULL;
        
        while(fast){
            fast = fast->next;
            slow = slow->next;
            if(!fast){
                return NULL;
            }
            fast = fast->next;
            if(fast == slow){
                meet = fast;
                break;
            }
        }
        while(meet && head){
            if(meet==head){
                return head;
            }
            meet = meet->next;
            head = head->next;
        }
        return NULL;
    }
};
